### DISCUSSION TITLE
fitting mods according to BPM

### DISCUSSION THREAD

**Discussion by Vilager58** [ppy/osu#28138] [Ideas] [answered]
Maybe the idea is not new and someone suggested it, but I did not find this discussion here.
### to the point:
make the bpm button in the mods menu clickable, and prompt the user to enter a new bpm, like in osu trainer.
![bpm](https://github.com/ppy/osu/assets/74284857/78764f66-3939-449a-a8b3-e7ec5702f0aa)
![bpm_edit](https://github.com/ppy/osu/assets/74284857/4040f459-56cd-44cd-af7c-09403ece4757)
and after the user enters a new bpm, the osu automatically configured the dt or ht mods to get this bpm

**Comment by Ekischleki** [comment]
Do you want the entered bpm to stay consistent between new beatmaps, or just keep the dt multiplier, for example:
You pick a beatmap with 120 bpm, speed it up to 150 bpm, meaning dt will be automatically configured to x1.25.
Now you switch to another beatmap with 150 bpm, do you want the dt multiplier to reconfigure to x1, or should it stay at x1.25?

**Reply by Vilager58** [reply]
I think recalculating multiplier for each beatmap is more useful. And I also have a question: if the beatmap has 90 bpm, and the user wants to speed it up by more than 2 times, like to 200, then this is going beyond the boundaries of dt...

**Reply by Ekischleki** [reply]
I don't think clicking on the bpm is a viable option, as it might be confusing, and not clear for the user on what to do. I would rather suggest adding a "set bpm-mod", which isn't restricted to dt boundaries.

**Reply by Vilager58** [reply]
Yeah, its also a good way

**Reply by Vilager58** [reply]
![image](https://github.com/ppy/osu/assets/74284857/7c65ef32-86c1-4e34-bc85-e41525b6a9ec)
something like this?
idk how to display value..

**Reply by Ekischleki** [reply]
I've tried a similar design, however I couldn't find out how to change the speed of the track according to the track's bpm because I can only access one at a time.

**Reply by Vilager58** [reply]
I tried using the Ibeatmapinfo.BPM interface element, which in theory should display the bpm of the current map, but the problem is that speedchange in dt\ht mods takes the value double instead of int, and I can't just specify speedchange = BPM /Ibeatmapinfo.bpm

**Reply by Ekischleki** [reply]
Yeah, that's my problem too

**Comment by Givikap120** [comment]
that's one of the my ideas when I initially made this display
so I agree with it

**Comment by cl8n** [comment]
commenting because I saw @Ekischleki working on this -- I think the design proposed in OP here is much better than adding settings to mods (DT/HT) or adding a new mod. it is conceptually not the same as a mod because it applies mods for you, and changes the mods between maps of differing BPM. would be great UX if every setting on that panel was editable, e.g. type in a CS and have DA mod automatically applied

I think the only extra thing I would add is a small checkbox/tick kind of thing in the area that decides if your BPM setting will be carried across multiple maps, because you may not actually want that, even though it's the main point of this issue.

oh and some kind of indicator if the BPM you typed in can't actually be reached with any mod settings, because it's too fast or too slow. in this case, I think it would be good if it still applied the fastest DT option or the slowest HT option

**Reply by Vilager58** [reply]
Yp, we made this partially; now we need some way to get current beatmap bpm, may be you know?

**Reply by Ekischleki** [reply, marked as answer]
@Vilager58 Members of the OSU team and I have come to the conclusion that we won't be implementing a mod, but rather doing something like @cl8n or you originally suggested, because making a mod would probably require significant changes to the mod backend, which isn't really wanted.

**Reply by Vilager58** [reply]
indeed, I've figured it out now, it turns out that in addition to transferring bpm, it will also need to take into info the unloading of the replay, interaction in multi...

**Reply by crumpledmint** [reply]
From a design standpoint you can just add the lock icon to the right of all of the parameters (🔒is parameters stay for the next map(s) 🔓is parameters will only apply for this map)
